Update the UI of the Juju Web CLI

hatched opened the pull request:
## Done

- Update the UI of the Juju Web CLI to the latest design.
- Removed the old `Terminal` component and dependencies.

## QA

- Bootstrap a new controller using the Juju 2.9 snap
  - `snap refresh --channel latest/beta`
- Build this PR and follow [this guide to connect to your controller](https://github.com/canonical-web-and-design/jaas-dashboard#developing-while-connected-to-a-juju-controller)
- View your controller model and the terminal should appear along the bottom.
- Steps:
  - - [ ] Click the help icon, did it show the help?
  - - [ ] Type `status` in the prompt and submit, did the model status appear?
  - - [ ] Type `status --color` in the prompt and submit, did the old status get cleared and a new coloured one appear?
  - - [ ] Type `help commands` in the prompt and submit, can you scroll the output
  - - [ ] Click and drag along the top of the output, does it resize as expected
  - - [ ] Drag it to the bottom so it's "closed"
  - - [ ] Type `status` in the prompt and submit, did it re-open?
  - - [ ] Click the help icon, did it resize and show the help content?
  - - [ ] Chaos monkey the interactions, can you find any bugs? Improvements?
  - - [ ] Chaos monkey the juju commands, can you find any bugs? Improvements?

[Zeplin Designs](https://app.zeplin.io/project/5f6234f4cb1e87aac4650d25/dashboard?sid=5f6235582ba7c6a777814c7f)

## Details

Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1445
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1446
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1447
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1448
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1456
Fixes https://github.com/canonical-web-and-design/juju-squad/issues/1396

## Screenshots
![Screen Shot 2020-10-21 at 12 07 20 PM](https://user-images.githubusercontent.com/532033/96788504-da259500-13b0-11eb-990f-87afaa1450b1.png)

webteam-app commented:
Demo starting at https://jaas-dashboard-720.demos.haus

anthonydillon commented on src/components/WebCLI/Output.js:
```diff
@@ -0,0 +1,144 @@
+  35: "188,63,188", // Magenta
+  36: "17,168,205", // Cyan
+  37: "229,229,229", // White
+  90: "102,102,102", // Bright Black (Gray)
```
Haha, Bright Black (Gray)

hatched replied:
Haha yeah I copied the description from the wikipedia article above, I also thought it was a funny inclusion :D

anthonydillon commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,90 @@
+  width: calc(100vw - #{$sidebar-collapsed});
+
+  &__input {
+    background-color: black;
```
`$color-x-dark;`

anthonydillon commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,90 @@
+    }
+
+    &-input[type="text"] {
+      background-color: black;
```
`$color-x-dark`

anthonydillon commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,90 @@
+      // https://github.com/canonical-web-and-design/vanilla-framework/pull/3370
+      // XXX The unquote will need to be removed with the next vanilla release
+      font-family: unquote($font-monospace);
+      height: 2.5rem;
```
Maybe set the height (2.5rem) to a variable so you can update the height from a single variable at the top of the file.

hatched replied:
done

anthonydillon commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,90 @@
+    padding: 0 1rem;
+
+    &-prompt {
+      color: #39b54a;
```
Can you make this color a variable at the top of the document

anthonydillon commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,90 @@
+  }
+
+  &__output {
+    background-color: #2c001e;
```
Would be good to have a nicely named variable for this as a setting at the top of this file

hatched replied:
done

anthonydillon reviewed (commented):
Looking awesome @hatch few suggestions in the code.

spencerbygraves commented:
@hatched thanks, it's looking good. I noticed a couple of things, hope they make sense.

- If I click help and then drag it closed again, clicking the help icon doesn't open it again
- After entering `help commands` I scrolled to the bottom. When I then drag the window down to close it again, the content scrolls back up. Is that expected?
- If I open help and drag the window up a bit, when I then type 'status' it doesn't resize the window to fit it all in, some is lost at the bottom.
- Could the window only open up as much as it needs to? After I enter `status` there is quite a bit of space under the text
- There is some inconsistent text positioning, the help text is closer to the top of the window than when you input a command. Could you try to match the design please.

![image](https://user-images.githubusercontent.com/9105999/96893542-b894de00-1482-11eb-8ec1-40d8297b42db.png)

- Could this menu use the [contextual menu pattern](https://vanillaframework.io/docs/examples/patterns/contextual-menu/default). Also, that menu doesn't display the other commands I entered after status.

![image](https://user-images.githubusercontent.com/9105999/96893906-1c1f0b80-1483-11eb-9dff-a1dbdae7913a.png)

hatched commented:
>     * If I click help and then drag it closed again, clicking the help icon doesn't open it again

This is a bug, will fix

>     * After entering `help commands` I scrolled to the bottom. When I then drag the window down to close it again, the content scrolls back up. Is that expected?

I'm not able to reproduce this, but I have seen it happen before, I'll see if I can find a repro case for the 'bounce to full screen' issue.

>     * If I open help and drag the window up a bit, when I then type 'status' it doesn't resize the window to fit it all in, some is lost at the bottom.

This was intentional as I treated the value you set it to to be more important than what we wanted to set it to.

>     * Could the window only open up as much as it needs to? After I enter `status` there is quite a bit of space under the text

I played around with this but wasn't able to create a pleasing and reliable experience within the timebox, I'll create an issue to see if we can do this properly in a follow-up.

>     * There is some inconsistent text positioning, the help text is closer to the top of the window than when you input a command. Could you try to match the design please.

So the spacing is actually the same but the responses from the backend sometimes have an extra newline - I've filed a bug report with them to resolve that. Good catch :)

>     * Could this menu use the [contextual menu pattern](https://vanillaframework.io/docs/examples/patterns/contextual-menu/default). Also, that menu doesn't display the other commands I entered after status.

As discussed offline this isn't part of the dashboard, maybe an autocomplete in your browser, I've disabled native autocomplete on the input to try and disable that.

hatched commented:
@spencerbygraves @clagom @anthonydillon this is ready for some more eyes. I'm still working on the tests but I believe I've addressed all the issues (and then some) or created follow-ups for the others.

barrymcgee commented on src/components/WebCLI/Output.js:
```diff
@@ -0,0 +1,147 @@
+  useEffect(() => {
+    const resize = (e) => {
+      const viewPortHeight = window.innerHeight;
+      const mousePosition = e.clientY + 40; // magic number
```
Why is it a magic number?

hatched replied:
It's what felt right when dragging things around.

spencerbygraves commented:
Thanks @hatched. My first two points above are fixed (open/close help and scrolling).

I still think number 3 is important. For example, If I click help and then straight after type `status`, I can't see the result without dragging the window up. I think this could be the first experience people have when using it.

That aside I'll give a design +1 as you're going to follow up on the spacing issue.

Cheers.

barrymcgee commented:
The bottom of the terminal needs bottom padding to account for the command prompt - `whoami` is obscured

<img width="961" alt="Screenshot 2020-10-23 at 09 45 35" src="https://user-images.githubusercontent.com/505570/96977203-a9f70700-1514-11eb-92eb-16517effa5d2.png">

barrymcgee commented on package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "juju-dashboard",
-  "version": "0.3.0",
```
This doesn't seem reflected in the yarn.lock file?

barrymcgee commented:
Clicking to the right of the terminal content can incrementally reduce the height of the terminal.

![Screen Recording 2020-10-23 at 09 56 29 am](https://user-images.githubusercontent.com/505570/96978390-37872680-1516-11eb-93ba-ec0f32769be0.gif)

barrymcgee commented:
Clicking whitespace under model details tables can maximise terminal unexpectedly

![Screen Recording 2020-10-23 at 10 02 16 am](https://user-images.githubusercontent.com/505570/96979587-11ae5180-1517-11eb-8529-0a8e3aa96159.gif)

barrymcgee commented on src/components/WebCLI/_webcli.scss:
```diff
@@ -0,0 +1,94 @@
+    pre {
+      background-color: $canonical-purple;
+      border: none;
+      color: white;
```
```suggestion
      color: $color-x-light;
```

barrymcgee commented on src/components/WebCLI/WebCLI.js:
```diff
@@ -0,0 +1,153 @@
+  const messageBuffer = useCallback(
+    (message) => {
+      wsMessageStore.current = wsMessageStore.current + message;
+      setTimeout(() => {
```
Doesn't this setTimeout require a timeout value?

hatched replied:
When you use setTimeout without a time it effectively pops it to the top of the stack to be executed on the next loop.

spencerbygraves commented:
One other thing, as you can drag the shell from any point, not just the drag bar, should the cursor be different?

barrymcgee commented on src/components/WebCLI/WebCLI.js:
```diff
@@ -0,0 +1,153 @@
+        <form onSubmit={handleCommandSubmit}>
+          <input
+            autoComplete="off"
+            className="webcli__input-input"
```
🤨

hatched replied:
it's an inputception

barrymcgee commented on src/components/WebCLI/Output.js:
```diff
@@ -0,0 +1,147 @@
+const DEFAULT_HEIGHT = 300;
+
+const WebCLIOutput = ({ content, helpMessage, showHelp }) => {
+  const resizeDeltaY = useRef(0);
```
AFAIK a `useRef` only ever returns an object so this default should be `null`, not `0`.

hatched replied:
It returns an object `{current: <initialValue>}` and the value passed to the `useRef()` sets that `initialValue`. So in this case I wanted the delta to start off as being 0.

barrymcgee commented on src/pages/Models/Details/ModelDetails.js:
```diff
@@ -376,7 +383,13 @@ const ModelDetails = () => {
+          credentials={credentials}
+          modelUUID={modelUUID}
+        />
+      ) : null}
```
If you have nothing to show as an alternative, you can simplify this conditional like so;

```
      {showWebCLI &&
        <WebCLI
          controllerWSHost={controllerWSHost}
          credentials={credentials}
          modelUUID={modelUUID}
        />
     }
```

hatched replied:
done

anthonydillon commented:
QA issue when trying to copy a public address:
![Peek 2020-10-23 10-54](https://user-images.githubusercontent.com/1413534/96989939-512c6c00-151e-11eb-8c7d-03113f43fd1b.gif)

hatched commented:
Sorry folks I pushed the wrong tarball on my way out the door last night, The one up there now has all of these issues resolved.

hatched commented on src/components/WebCLI/Output.js:
```diff
@@ -0,0 +1,147 @@
+      const mousePosition = e.clientY + 40; // magic number
+      const newHeight = viewPortHeight - mousePosition + resizeDeltaY.current;
+
+      const maximumOutputHeight = viewPortHeight - 100; // magic number.
```
It's just what felt right spacing wise

hatched commented:
OK This one is now all ready to go, QA environment has been updated with the latest build.

anthonydillon commented:
Thanks @hatched this is looking great. I am testing on my phone and here are a few things I found. None are blockers so feel free to fix in a follow up.

- Add `autocapitalize="off"` to the input__input. (Sorry, in joke). It's annoying for phones to capitalise commands by default
- Can you add an analytics event to command. I think that will help us make decisions on what users want.
- Can you add touch events to the resizing code as I can't resize the CLI on my phone

hatched commented:
Thanks for the input @anthonydillon I've resolved the first two, but created an issue for the last one as it appears to be a bit bigger of an issue.

ziheliu214 commented:
@hatched Thanks Jeff amazing work!

UX questions:

1) Was the output designed to be one-off record, that every time jumping to different pages (including clicking on the help) it would be erased and not be able to scroll to historical output?

2)
![image](https://user-images.githubusercontent.com/61144986/97202135-20ad3200-17ab-11eb-86ad-2dd2c13364cd.png)
When a wrong command is entered, it suggests to try `run` but I am not sure if it is supported.

hatched commented:
Thanks @ziheliu214 
1. Yes the output intentionally clears after every new command submission. This was so that it was more clear that this isn't a 'shell' but a way to run juju cli commands. We can adjust this easily so lets wait to hear back from users what they think.
2. It looks like you submitted an empty command, We should probably prevent that. The output you're seeing is directly from Juju. It's trying to hint to you what command you maybe meant to run.

anthonydillon reviewed (approved):
Code and QA :+1: great work @hatched

hatched commented:
Thanks all for the reviews and qa's!
